utils/src/foreign_field.rs

```diff
@@ -152,48 +138,379 @@ impl<F: Field, const N: usize> IndexMut<usize> for ForeignElement<F, N> {
     }
 }
 
+impl<F: Field, const N: usize> Debug for ForeignElement<F, N> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "ForeignElement(")?;
+        for i in 0..self.len {
+            write!(f, "{:?}", self.limbs[i].to_hex())?;
+            if i != self.len - 1 {
+                write!(f, ", ")?;
+            }
+        }
+        write!(f, ")")
+    }
+}
+
+/// Foreign field helpers
+pub trait ForeignFieldHelpers<T> {
+    /// 2^{LIMB_BITS}
+    fn two_to_limb() -> T;
+
+    /// 2^{2 * LIMB_BITS}
+    fn two_to_2limb() -> T;
+}
+
+impl<F: Field> ForeignFieldHelpers<F> for F {
+    fn two_to_limb() -> Self {
+        F::from(2u64).pow(&[LIMB_BITS as u64])
+    }
+
+    fn two_to_2limb() -> Self {
+        F::from(2u64).pow(&[2 * LIMB_BITS as u64])
+    }
+}
+
+/// Foreign field helpers
+pub trait BigUintForeignFieldHelpers {
+    /// 2
+    fn two() -> Self;
+
+    /// 2^{LIMB_SIZE}
+    fn two_to_limb() -> Self;
+
+    /// 2^{2 * LIMB_SIZE}
+    fn two_to_2limb() -> Self;
+
+    /// 2^t
+    fn binary_modulus() -> Self;
+
+    /// Convert to 3 limbs of LIMB_BITS each
+    fn to_limbs(&self) -> [BigUint; 3];
+
+    /// Convert to 2 limbs of 2 * LIMB_BITS each
+    fn to_compact_limbs(&self) -> [BigUint; 2];
+
+    /// Convert to 3 PrimeField limbs of LIMB_BITS each
+    fn to_field_limbs<F: Field>(&self) -> [F; 3];
+
+    /// Convert to 2 PrimeField limbs of 2 * LIMB_BITS each
+    fn to_compact_field_limbs<F: Field>(&self) -> [F; 2];
+
+    /// Negate: 2^T - self
+    fn negate(&self) -> BigUint;
+}
+
+impl BigUintForeignFieldHelpers for BigUint {
+    fn two() -> Self {
+        Self::from(2u32)
+    }
+
+    fn two_to_limb() -> Self {
+        BigUint::two().pow(LIMB_BITS as u32)
+    }
+
+    fn two_to_2limb() -> Self {
+        BigUint::two().pow(2 * LIMB_BITS as u32)
+    }
+
+    fn binary_modulus() -> Self {
+        BigUint::two().pow(3 * LIMB_BITS as u32)
+    }
+
+    fn to_limbs(&self) -> [Self; 3] {
+        let mut limbs = biguint_to_limbs(self, LIMB_BITS);
+        assert!(limbs.len() <= 3);
+        limbs.resize(3, BigUint::zero());
+
+        array::from_fn(|i| limbs[i].clone())
+    }
+
+    fn to_compact_limbs(&self) -> [Self; 2] {
+        let mut limbs = biguint_to_limbs(self, 2 * LIMB_BITS);
+        assert!(limbs.len() <= 2);
+        limbs.resize(2, BigUint::zero());
+
+        array::from_fn(|i| limbs[i].clone())
+    }
+
+    fn to_field_limbs<F: Field>(&self) -> [F; 3] {
+        self.to_limbs().to_field_limbs()
+    }
+
+    fn to_compact_field_limbs<F: Field>(&self) -> [F; 2] {
+        self.to_compact_limbs().to_field_limbs()
+    }
+
+    fn negate(&self) -> BigUint {
```

Comment:
Not sure you only want this function... This is okay for computing the negated modulus, but if you ever want to negate any other value within the foreign field modulus, I think we may also want this. Meaning, computing -3 for example, but in secp256, not in the binary modulus.

Reply:
Ah, nevermind, `neg()` is still there. Well, perhaps the wording is not the best to differentiate between the two, but ok